api/manageiq-api.md

```diff
@@ -0,0 +1,217 @@
+ManageIQ API (Developer Guide)
+==============================
+
+This is a developer guide for navigating the `manageiq-api` repo, and how to
+understand the codebase to make additions and fixes yourself.  This guide will
+not go in depth into the functionality of each resource, but rather explain
+common functionality across all controllers to allow you to determine nuances
+on a per-resource basis for yourself.
+
+A good introduction (that I found after writing almost half of this...) was
+written by Jillian Tullo:
+
+http://manageiq.org/blog/2017/10/api-contributors-guide/
+
+And is a great overview of what is talked about this guide.
+
+That said, this is also meant to be a living document, not a point in time
+reference, so there at least should still be some value in this (I hope...
+:sob:)
+
+
+Table of Contents
+-----------------
+
+* [The `base_controller.rb`](#the-base_controllerrb)
+  - [The Basics](#the-basics)
+  - [Overriding base functionality](#overriding-base-functionality)
+  - [The `RequestAdapter`](#the-requestadapter)
+* [ApiConfig](#apiconfig)
+  - [The `config/api.yml`](#the-configapiyml)
+    - [`:options:`](#options)
+      - [Collections and Subcollections](#collections-and-subcollections)
+      - [Validate Action](#validate-action-validate_action)
+      - [Custom Actions](#custom-actions-custom_actions)
+  - [Drawing Routes (the `config/routes.rb`)](#drawing-routes-the-configroutesrb)
+* [The `manageiq-api-client` gem](#the-manageiq-api-client-gem)
+
+
+The `base_controller.rb`
+------------------------
+
+- Most Controllers inherit from this controller
+- Updating resource options
+
+The first part of `manageiq-api` codebase that you should familiarize yourself
+with is the [`app/controllers/api/base_controller.rb`][1] class, and included
+modules in it's associated directory.  Almost all (if not all), controllers in
+the API inherit from this controller, and the entry point for each action
+defined in the routes (more on that in a later section), are defined in this
+class
+
+
+### The Basics
+
+If you are familiar with a typical Rails controller, most of the
+methods/actions you are used to seeing are there:
+
+- `index`
+- `show`
+- `create`
+- `update`
+- `destroy`
+
+And so on, but most of these calls are inherited and never modified.  So for
+example, the controller for accessing all of the `MiqServer` records in an
+installation of ManageIQ is defined in the `Api::ServersController`:
+
+[`app/controllers/api/servers_controller.rb`][2]
+
+But, that class itself is basically empty:
+
+```ruby
+module Api
+  class ServersController < BaseController
+  end
+end
+```
+
+This is because it uses, without modifications, the code from the
+`Api::BaseController`.  Now this is basic inheritance, so there is nothing
+interesting here of note, accept for the fact that not every method defined in
+the `Api::BaseController` is available as a route in the
+`Api::ServersController`:
+
```

Comment:
@NickLaMuro I think before showing the user of `bin/rake routes` it would be valuable to mention that this command must be run in the manageiq repo. See: https://github.com/ManageIQ/manageiq-api#development

Perhaps:
Development of the manageiq-api code base is done by pointing your local clone of the manageiq repo
and pointing it to your local branch manageiq-api repo. How to do this is described in [this section ](https://github.com/ManageIQ/guides/blob/master/developer_setup/plugins.md#dependency-on-a-local-gem) of the general developer setup guide.

Once this is set up it will be possible to query the established routes from by running the following commands from the top level directory of your manageiq repository.